### PR TITLE
Halo RTD Module: FPD 2.0 Updates & add ID system tests

### PR DESCRIPTION
**Pull request by anthonylauzon**
## Type of change
Feature

## Description of change
This changeset updates our RTD module to allow ORTB2 data to be retrieved and merged with the setConfig ORTB2 data. Examples of this type of data include IAB taxonomy segments.  The changeset additionally simplifies how publishers can use Halo real-time data in other use cases and adds halo ID system unit tests.

## Contact
ant@audigent.com

## Other information

Fpd 2.0 Update https://github.com/prebid/Prebid.js/pull/6293
First Party Data Module https://github.com/prebid/Prebid.js/issues/6099
First Party Data Revision https://github.com/prebid/Prebid.js/issues/5795

### CODE REVIEW

**Comment by anthonylauzon** [comment]
@bretg @patmmccann  this changeset should be ready to go, but would appreciate any feedback as usual.  looking forward to an e2e test integration with the magnite adapter and one of our pubs once this is cleared.

**Comment by bretg** [comment]
@anthonylauzon - I looked through the code and examples some and am not understanding how this will work for the POC. Can you post what the config will look like?

I had been expecting to see "taxonomyname" in the code -- assumed that your endpoint would return a list of segments and the module would create a user.data array entry, read in what's in ortb2, add the new entry, the write back ortb2.

Instead, the test cases are doing something not very useful-seeming. And something I wouldn't recommend in production -- the example rtdHandler is doing a setconfig on ortb2 without first doing a getconfig and merge.

**Comment by anthonylauzon** [comment]
@bretg the first test case is the POC test case, and it only requires a publisher id: 

`    realTimeData: {
        auctionDelay: auctionDelay,
        dataProviders: [
            {
                name: "halo",
                waitForIt: true,
                params: {
                    segmentCache: false,
                    requestParams: {
                        publisherId: 1234
                    }
                }
            }
        ]
    }`

Our backends will return ortb2 compatible data and merge it into any other data present in setConfig.  For the POC, this will be IAB taxonomy segments as you can see in the first test case, which we use different data provider labels for just for the purposes of testing.  We won't be merging segments based on taxonomy name, since taxonomies are differentiated by data provider.  This was discussed during one of the POC meetings.

The handleRtd function's test case is just a test and not something that you would see in production.  A better example of what handleRtd might look like in production is in the documentation.  It is provided as both a backwards compatibility layer for RTD ad-unit level logic and to allow other types of integrations outside of the scope of the POC.  This changeset continues to allow for existing ad-unit and bid adapter specific logic which while also enabling IAB segment poc testing.

**Comment by bretg** [comment]
Ok, having the formatting on your server-side may make sense given that there may be some syntax changes coming.  I'll let the POC team work through any outstanding details here.

**Comment by patmmccann** [comment]
> Our backends will return ortb2 compatible data and merge it into any other data present in setConfig.

This makes sense to me as well, ty

**Comment by anthonylauzon** [comment]
https://github.com/prebid/Prebid.js/issues/6057 potential format changes are handled here

**Comment by robertrmartinez** [comment]
Sorry for the delay.

I will do this review today!

**Comment by anthonylauzon** [comment]
> Sorry for the delay.
> 
> I will do this review today!

thank you & no problem @robertrmartinez

File: modules/haloRtdProvider.js [hunk 1]
```diff
@@ -1,109 +1,102 @@
...
+  if (!isPlainObject(config.getConfig().ortb2)) {
+    config.setConfig({ortb2: {}});
+  }
 
-  return adUnits;
+  if (rtdConfig.params && rtdConfig.params.handleRtd) {
+    rtdConfig.params.handleRtd(bidConfig, rtd, rtdConfig, config);
+  } else if (rtd.ortb2) {
+    let mergedOrtb = mergeLazy(config.getConfig().ortb2, rtd.ortb2);
+    config.setConfig({ortb2: mergedOrtb});
```
[the remark this answers is not in the document]
Comment (robertrmartinez) [reply modules/haloRtdProvider.js:80]: Don't think it makes sense to setConfig on a empty ortb2 object like line 73

File: modules/haloRtdProvider.js [hunk 2]
```diff
@@ -1,109 +1,102 @@
...
-  const adUnits = reqBidsConfigObj.adUnits || getGlobal().adUnits;
-
-  if (config.params.segmentCache) {
-    let jsonData = storage.getDataFromLocalStorage(SEG_LOCAL_NAME);
+export function getRealTimeData(bidConfig, onDone, rtdConfig, userConsent) {
+  if (isPlainObject(rtdConfig.params) && rtdConfig.params.segmentCache) {
+    let jsonData = storage.getDataFromLocalStorage(RTD_LOCAL_NAME);
 
     if (jsonData) {
       let data = JSON.parse(jsonData);
```
Comment (robertrmartinez) [inline modules/haloRtdProvider.js:96]: Maybe surround in try catch just in case? Outside of this PR change so feel free to ignore

**Review by robertrmartinez — CHANGES_REQUESTED** [review]
couple questions / comments

File: modules/haloIdSystem.js [hunk 3]
```diff
@@ -5,12 +5,15 @@
...
 
-import * as utils from '../src/utils.js';
 import {ajax} from '../src/ajax.js';
+import {getStorageManager} from '../src/storageManager.js';
 import {submodule} from '../src/hook.js';
+import * as utils from '../src/utils.js';
 
 const MODULE_NAME = 'haloId';
 
+export const storage = getStorageManager();
```
[the remark this answers is not in the document]
Comment (anthonylauzon) [reply modules/haloIdSystem.js]: added to both modules

File: modules/haloRtdProvider.js [hunk 4]
```diff
@@ -1,109 +1,102 @@
...
+import {config} from '../src/config.js';
 import {getGlobal} from '../src/prebidGlobal.js';
-import * as utils from '../src/utils.js';
+import {getStorageManager} from '../src/storageManager.js';
 import {submodule} from '../src/hook.js';
-import {ajax} from '../src/ajax.js';
-import { getStorageManager } from '../src/storageManager.js';
+import {isFn, isStr, isPlainObject, mergeDeep, logError} from '../src/utils.js';
 
 export const storage = getStorageManager();
```
[the remark this answers is not in the document]
Comment (anthonylauzon) [reply modules/haloRtdProvider.js]: added to both modules

File: modules/haloRtdProvider.js [hunk 5]
```diff
@@ -69,15 +70,12 @@ function paramOrDefault(param, defaultVal) {
  * @param {Object} rtdConfig
  */
 export function addRealTimeData(bidConfig, rtd, rtdConfig) {
-  if (!isPlainObject(config.getConfig().ortb2)) {
-    config.setConfig({ortb2: {}});
-  }
+  let ortb2 = config.getConfig().ortb2 || {};
```
Comment (robertrmartinez) [inline modules/haloRtdProvider.js:73]: if you pass in the specific config to getConfig you want, it will only return that, instead of the entire configs which have been set.

`config.getConfig('ortb2');`

This is recommended because If for some reason `config.getConfig()` does not return an object, then `config.getConfig().ortb2` will throw an exception.

Comment (anthonylauzon) [reply modules/haloRtdProvider.js]: thanks, fixed

File: test/spec/modules/haloIdSystem_spec.js [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+      expect(request.url).to.eq(`https://id.halo.ad.gt/api/v1/pbhid`);
+      request.respond(200, { 'Content-Type': 'application/json' }, JSON.stringify({ haloId: 'testHaloId1' }));
+      expect(callbackSpy.lastCall.lastArg).to.deep.equal({haloId: 'testHaloId1'});
+    });
+
+    it('gets a cached haloid', function() {
+      const config = {
+        params: {}
+      };
+      storage.setDataInLocalStorage('auHaloId', 'tstCachedHaloId1');
```
Comment (robertrmartinez) [inline test/spec/modules/haloIdSystem_spec.js:24]: It is not recommended to actually have the tests themselves setting localStorage / cookies.

These can become flakey depending on which browsers they are run in.

See https://github.com/prebid/Prebid.js/blob/master/test/spec/modules/rubiconAnalyticsAdapter_spec.js#L585-L589 as an example of mocking your storage object.

Then you can mock it's return like so: https://github.com/prebid/Prebid.js/blob/master/test/spec/modules/rubiconAnalyticsAdapter_spec.js#L1250


Also do not forget to call restore() on each stub in the afterEach's => https://github.com/prebid/Prebid.js/blob/master/test/spec/modules/rubiconAnalyticsAdapter_spec.js#L618-L620

File: test/spec/modules/haloRtdProvider_spec.js [hunk 7]
```diff
@@ -1,220 +1,363 @@
...
+          ortb2: {
+            user: {
+              data: [rtdUserObj1]
+            }
+          }
+        }
+      };
 
-      getSegments(reqBidsConfigObj, () => {}, config, {});
+      storage.setDataInLocalStorage(RTD_LOCAL_NAME, JSON.stringify(cachedRtd));
```
Comment (robertrmartinez) [inline test/spec/modules/haloRtdProvider_spec.js:279]: Same thing here, stub this.

File: test/spec/modules/haloIdSystem_spec.js (hunk 6 above)
Comment (anthonylauzon) [reply test/spec/modules/haloIdSystem_spec.js:24]: i hadn't seen this pattern before, thanks for the info.

**Review by robertrmartinez — APPROVED** [review]
Looks good to me! 

Nice work, bit of a doosy this one!

**Comment by robertrmartinez** [comment]
@ChrisHuie When you have some time, please give this a second look! Thanks.

**Review by ChrisHuie — APPROVED** [review]
